### PR TITLE
Add support for launching a browser on file change

### PR DESCRIPTION
**Pull request by pranavkm**
Contributes to https://github.com/dotnet/aspnetcore/issues/23412

### CODE REVIEW

File: src/Tools/dotnet-watch/src/BrowserRefreshServer.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            {
+                context.Response.StatusCode = 400;
+                return;
+            }
+
+            _webSocket = await context.WebSockets.AcceptWebSocketAsync();
+            await _taskCompletionSource.Task;
+        }
+
+        public async void SendMessage(byte[] messageBytes)
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/BrowserRefreshServer.cs:106]: 😭 Task

File: src/Tools/dotnet-watch/src/BrowserRefreshServer.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                .GetRequiredService<IServer>()
+                .Features
+                .Get<IServerAddressesFeature>()
+                .Addresses
+                .First();
+
+            return serverUrl.Replace("http://", "ws://");
+        }
+
+        static void RunInBackgroundThread(IHost host)
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/BrowserRefreshServer.cs:64]: What is the point of this?

File: src/Tools/dotnet-watch/src/BrowserRefreshServer.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        private WebSocket _webSocket;
+
+        public BrowserRefreshServer(
+            IReporter reporter)
+        {
+            _reporter = reporter;
+            _taskCompletionSource = new TaskCompletionSource();
+        }
+
+        public string Start()
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/BrowserRefreshServer.cs:36]: Can we convert everything to use async Task? `host.StartAsync`, `host.StopAsync`, `IAsyncDisposable` etc.

File: src/Tools/dotnet-watch/src/BrowserRefreshServer.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                await _webSocket.SendAsync(messageBytes, WebSocketMessageType.Text, endOfMessage: true, CancellationToken.None);
+            }
+            catch (Exception ex)
+            {
+                _reporter.Output($"Refresh server error: {ex}");
+            }
+        }
+
+        public void Dispose()
+        {
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/BrowserRefreshServer.cs:124]: Probably should be calling `CloseOutputAsync` on the websocket here.

File: src/Tools/dotnet-watch/src/BrowserRefreshServer.cs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            _refreshServer?.Dispose();
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/BrowserRefreshServer.cs:125]: We should be calling Stop/StopAsync on `IHost`

File: src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+using Microsoft.Extensions.Tools.Internal;
+
+namespace Microsoft.DotNet.Watcher.Tools
+{
+    public sealed class LaunchBrowserFilter : IWatchFilter, IDisposable
+    {
+        private readonly byte[] ReloadMessage = Encoding.UTF8.GetBytes("Reload");
+        private readonly byte[] WaitMessage = Encoding.UTF8.GetBytes("Wait");
+        private static readonly Regex NowListeningRegex = new Regex(@"^\s*Now listening on: (?<url>.*)$", RegexOptions.None, TimeSpan.FromSeconds(10));
+        private static readonly Regex ApplicationStartedRegex = new Regex(@"^\s*Application started\. Press Ctrl\+C to shut down\.$", RegexOptions.None, TimeSpan.FromSeconds(10));
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs:23]: Compiled regex?

File: src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+            var args = string.Empty;
+            if (!string.IsNullOrEmpty(_browserPath))
+            {
+                args = fileName;
+                fileName = _browserPath;
+            }
+
+            if (_runningInTest)
+            {
+                _reporter.Verbose($"Launching browser: {fileName} {args}");
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs:138]: Should this report be logged outside of tests too?

File: src/Tools/dotnet-watch/src/dotnet-watch.csproj [hunk 8]
```diff
@@ -19,4 +19,24 @@
     <None Include="assets\**\*" CopyToOutputDirectory="PreserveNewest" CopyToPublishDirectory="PreserveNewest" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Reference Include="Microsoft.AspNetCore" />
+    <Reference Include="Microsoft.AspNetCore.WebSockets" />
+  </ItemGroup>
+
+  <Target Name="_FixupRuntimeConfig" BeforeTargets="_GenerateRuntimeConfigurationFilesInputCache">
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/dotnet-watch.csproj:27]: Idk about this stuff 😄

File: src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+                    }
+                }
+            }
+        }
+
+        private void LaunchBrowser(string launchUrl)
+        {
+            var fileName = launchUrl + "/" + _launchPath;
+            var args = string.Empty;
+            if (!string.IsNullOrEmpty(_browserPath))
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs:130]: Should there be an error here? If you didn't set a browser path it'll try to run "http://localhost:54304/something" as a command which I don't think works

File: src/Tools/dotnet-watch/src/BrowserRefreshServer.cs [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+            if (edi != null)
+            {
+                throw edi.SourceException;
+            }
+        }
+
+        async Task WebSocketRequest(HttpContext context)
+        {
+            if (!context.WebSockets.IsWebSocketRequest)
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/BrowserRefreshServer.cs:96]: Who is attaching with a WebSocket connection? (The browser I know, but how?)

Comment (pranavkm) [reply src/Tools/dotnet-watch/src/BrowserRefreshServer.cs:95]: There's a follow up which will inject a hosting startup via environment variable.

**Comment by pranavkm** [comment]
Updated

File: src/Tools/dotnet-watch/src/BrowserRefreshServer.cs (hunk 1 above)
Comment (pranavkm) [reply src/Tools/dotnet-watch/src/BrowserRefreshServer.cs:105]: The callsite is sync. I'm not sure if if helps to make this async and block it.

Comment (BrennanConroy) [reply src/Tools/dotnet-watch/src/BrowserRefreshServer.cs:105]: One of the callsites is async, and you can do `_ = SendMessage(...);` in the sync case.

File: src/Tools/dotnet-watch/src/BrowserRefreshServer.cs [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+            {
+                _reporter.Output($"Refresh server error: {ex}");
+            }
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            if (_webSocket != null)
+            {
+                await _webSocket.CloseOutputAsync(WebSocketCloseStatus.Empty, null, default);
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/BrowserRefreshServer.cs:126]:
```suggestion
                await _webSocket.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, null, default);
```

**Comment by pranavkm** [comment]
Good to go

File: src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+                {
+                    // We've detected a change. Notify the browser.
+                    await _refreshServer.SendMessage(WaitMessage, cancellationToken);
+                }
+            }
+        }
+
+        private void OnOutput(object sender, DataReceivedEventArgs eventArgs)
+        {
+            // We've redirected the output, but want to ensure that continues to appear in the user's console.
```
Comment (pranavkm) [inline src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs:83]:
```suggestion
            // We've redirected the output, but want to ensure that it continues to appear in the user's console.
```

File: src/Tools/dotnet-watch/test/BrowserLaunchTests.cs [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+            await _app.Process.GetOutputLineStartsWithAsync(launchBrowserMessage, TimeSpan.FromMinutes(2));
+
+            // Make a file change and verify we reloaded the browser.
+            File.SetLastWriteTime(source, DateTime.Now);
+            await _app.Process.GetOutputLineStartsWithAsync(refreshBrowserMessage, TimeSpan.FromMinutes(2));
+        }
+
+        [ConditionalFact]
+        [OSSkipCondition(OperatingSystems.Linux)]
+        public async Task UsesBrowserSpecifiedinEnvironment()
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/test/BrowserLaunchTests.cs:55]:
```suggestion
        public async Task UsesBrowserSpecifiedInEnvironment()
```

File: src/Tools/dotnet-watch/src/Internal/FileSet.cs [hunk 14]
```diff
@@ -11,16 +11,19 @@ public class FileSet : IFileSet
...
+            IsNetCoreApp31OrNewer = isNetCoreApp31OrNewer;
             _files = new HashSet<string>(files, StringComparer.OrdinalIgnoreCase);
         }
 
         public bool Contains(string filePath) => _files.Contains(filePath);
 
         public int Count => _files.Count;
 
-        public static IFileSet Empty = new FileSet(Array.Empty<string>());
+        public bool IsNetCoreApp31OrNewer { get; }
```
Comment (BrennanConroy) [inline src/Tools/dotnet-watch/src/Internal/FileSet.cs:24]: What is stopping 2.1 from working?

Comment (pranavkm) [reply src/Tools/dotnet-watch/src/Internal/FileSet.cs:24]: The HTML injection middleware has to target whatever version we choose here. We could have it just launch the browser the first time, but that's not very useful in itself. Picking 3.1 seems like a good middleground.

File: src/Tools/dotnet-watch/src/dotnet-watch.csproj [hunk 15]
```diff
@@ -19,4 +19,24 @@
...
+    <ItemGroup>
+      <_RuntimeFramework Include="@(RuntimeFramework)" />
+      <RuntimeFramework Remove="@(RuntimeFramework)" />
+      <RuntimeFramework Include="Microsoft.AspNetCore.App" FrameworkName="Microsoft.AspNetCore.App" Version="5.0.0-preview" />
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Tools/dotnet-watch/src/dotnet-watch.csproj:31]: I'm on point for fixing this in 5.0. We basically want to stamp the current runtime version in here. This was copied from https://github.com/dotnet/aspnetcore/blob/master/src/Components/WebAssembly/DevServer/src/Microsoft.AspNetCore.Components.WebAssembly.DevServer.csproj#L41 so we have a few places to fix this.

File: src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+                process.OutputDataReceived -= OnOutput;
+                process.CancelOutputRead();
+
+                if (!_browserLaunched)
+                {
+                    _reporter.Verbose("Launching browser.");
+                    _browserLaunched = true;
+                    try
+                    {
+                        LaunchBrowser(launchUrl);
```
Comment (pranavkm) [inline src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs:106]:
```suggestion
                        LaunchBrowser(launchUrl);
                        _browserLaunched = true;
```

File: src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+                var launchUrl = match.Groups["url"].Value;
+
+                var process = (Process)sender;
+                process.OutputDataReceived -= OnOutput;
+                process.CancelOutputRead();
+
+                if (!_browserLaunched)
+                {
+                    _reporter.Verbose("Launching browser.");
+                    _browserLaunched = true;
```
Comment (pranavkm) [inline src/Tools/dotnet-watch/src/LaunchBrowserFilter.cs:103]:
```suggestion
```

**Comment by Tratcher** [comment]
Bad rebase?

**Comment by pranavkm** [comment]
Wrong base branch.

**Review by BrennanConroy — APPROVED** [review]

**Pull request merged**